### DISCUSSION TITLE
How do correctly describe the opacity for a colors?

### DISCUSSION THREAD

**Discussion by afuno** [tailwindlabs/tailwindcss#2206] [Help]
I'm using version 1.7.2.

There is an example on the release page:

```js
// tailwind.config.js
module.exports = {
  theme: {
    colors: {
      primary: ({ opacityVariable }) => `rgba(var(--color-primary), var(${variable}, 1))`,
    },
  },
}
```

In this code, there is an error in the variable.

This example has a link to Pull (https://github.com/tailwindlabs/tailwindcss/pull/1676) which has this code:

```js
// tailwind.config.js
module.exports = {
  theme: {
    colors: {
      primary: variable => `rgba(32, 32, 32, var(${variable}, 1))`,
    },
  },
}
```

Options using only `opacityVariable` or only `variable` do not work for me. I always get the error:

> `@apply` cannot be used with `.border-primary` because `.border-primary` either cannot be found, or its actual definition includes a pseudo-selector like :hover, :active, etc. If you're sure that `.border-primary` exists, make sure that any `@import` statements are being properly processed *before* Tailwind CSS sees your CSS, as `@apply` can only be used for classes in the same CSS tree.

When I return everything to my original code, then everything is fine:

```js
primary: 'var(--color-primary)',
```

My non-working options based on your recommendations:

```js
primary: ({ opacityVariable }) => `rgba(var(--color-primary), var(${opacityVariable}, 1))`,
// or
primary: ({ variable }) => `rgba(var(--color-primary), var(${variable}, 1))`,
```

Can you please tell me what I'm doing wrong?

**Comment by adamwathan** [comment]
This should be improved in v1.7.3, there was a bug (although I think it was a different bug than what you're seeing).

I'd recommend defining closure-based colors like this:

```js
primary: ({ opacityValue, opacityVariable }) => {
  if (opacityValue !== undefined) {
    return `rgba(var(--color-primary), ${opacityValue})`
  }

  if (opacityVariable !== undefined) {
    return `rgba(var(--color-primary), var(${opacityVariable}, 1))`
  }
  
  return `rgb(var(--color-primary))`
},
```

See related discussion here: https://github.com/tailwindlabs/tailwindcss/pull/2185

**Comment by afuno** [comment]
@adamwathan I copied and pasted this code and now I get this error:

> TypeError: value.charCodeAt is not a function

**Comment by afuno** [comment]
And this code does not work for me:

```js
primary: 'rgb(var(--color-primary))',
```

It is defined in the inspector in the browser. And there is even a color icon. That is, in the inspector in the browser, the color seems to have been determined.

<img width="278" alt="Снимок экрана 2020-08-20 в 18 17 01" src="https://user-images.githubusercontent.com/2853419/90791012-63690e80-e311-11ea-9c54-771d9e3533ed.png">

But on the pages, the text is not painted over. Only the default color from the body.

**Comment by adamwathan** [comment]
Can you share a project that reproduces the issue please? We usually immediately close issues that don't contain a reproduction.

**Comment by afuno** [comment]
@adamwathan Unfortunately I can not. I just don't have one in the compiled code like in your documentation:

```css
.text-purple-600 {
    --text-opacity: 1!important;
    color: #805ad5!important;
    color: rgba(128,90,213,var(--text-opacity)))!important;
}
```

For some reason, mine is like in the screenshot above.

I added this default code to my config file:

```js
purple: {
  100: '#faf5ff',
  200: '#e9d8fd',
  300: '#d6bcfa',
  400: '#b794f4',
  500: '#9f7aea',
  600: '#805ad5',
  700: '#6b46c1',
  800: '#553c9a',
  900: '#44337a',
},
```

And for some reason, more properties appear in the compiled code:

```css
.text-purple-100 {
    --text-opacity: 1;
    color: #faf5ff;
    color: rgba(250, 245, 255, var(--text-opacity));
}
```

But I don't know whether these problems are related or not.

**Comment by afuno** [comment]
@adamwathan I only kept one color:

```js
colors: {
  primary: ({ opacityValue, opacityVariable }) => {
    return `rgb(var(--color-primary))`
  },
},
```

> TypeError: value.charCodeAt is not a function

**Comment by afuno** [comment]
For:

```js
primary: '#ff0000',
```

It turns out this:

```css
.text-primary {
    --text-opacity: 1;
    color: #ff0000;
    color: rgba(255, 0, 0, var(--text-opacity));
}
```

For:

```js
primary: 'var(--color-primary)',
```

It turns out this:

```css
.text-primary {
    color: var(--color-primary);
}
```

**Comment by adamwathan** [comment]
Can't help without a reproduction unfortunately. Please create a new GitHub repository that demonstrates the problem and provide a link so we can help. It doesn't have to be your real project, just a new demo project that actually shows the bug. Very good chance there is no bug here and just misconfiguration or a misunderstanding of how things are meant to work.

**Comment by vseventer** [comment]
@afuno @adamwathan I ran into the same issue, and found out that, as far as I can tell, the color variable needs to be declared in RGB format for all of this to work.

* Doesn't work: `--color-primary: #ff0000;`
* Works: `--color-primary: 255, 0, 0;`

**Reply by afuno** [reply]
@vseventer  I apologize for the long answer. For me, all three options give the same error:

- `--color-primary: #ff0000;`
- `--color-primary: 255, 0, 0;`
- `--color-primary: rgb(255, 0, 0);`

**Comment by afuno** [comment]
And I don't understand why, for example, in the documentation, the text has these styles:

<img width="300" alt="Снимок экрана 2020-09-08 в 14 27 33" src="https://user-images.githubusercontent.com/2853419/92470795-a36c3480-f1df-11ea-8717-7c3ab9688da8.png">

But my text only has this:

<img width="300" alt="Снимок экрана 2020-09-08 в 14 28 21" src="https://user-images.githubusercontent.com/2853419/92470866-bbdc4f00-f1df-11ea-8c25-dbbdee671c6b.png">

But if I replace this code:

```js
primary: 'var(--color-primary)',
```

On this:

```js
primary: '#ff0000',
// or
// primary: 'rgb(255, 0, 0)',
```

But new attributes appear:

<img width="300" alt="Снимок экрана 2020-09-08 в 14 31 19" src="https://user-images.githubusercontent.com/2853419/92471117-0e1d7000-f1e0-11ea-8a08-d08fd4dc6fcf.png">

But if we return to the problem, then, for example, this code:

```js
primary: ({ opacityValue, opacityVariable }) => {
  if (opacityValue !== undefined) {
    return `rgba(255, 0, 0, ${opacityValue})`
  }

  if (opacityVariable !== undefined) {
    return `rgba(255, 0, 0, var(${opacityVariable}, 1))`
  }

  return `rgb(255, 0, 0)`
},
```

Will still throw an error:

> TypeError: value.charCodeAt is not a function

**Reply by afuno** [reply]
@adamwathan Are there any options?

**Comment by afuno** [comment]
@adamwathan Does it have something to do with this https://github.com/tailwindlabs/tailwindcss/pull/2515?

**Reply by adamwathan** [reply]
Try out the latest release and see if it fixes your issue. I made a few improvements.

**Reply by afuno** [reply]
@adamwathan 

This code still doesn't work:

```js
primary: 'var(--color-primary)',
```

A similar option for RGBA doesn't work either.

The code above doesn't add this option either:

```css
--text-opacity: 1;
```

This code:

```js
primary: () => `rgb(var(--color-primary))`,
```

Generates this:

```css
--text-opacity: 1;
color: rgb(var(--color-primary));
```

Option `--text-opacity: 1` appeared but color still doesn't work.

**Reply by afuno** [reply]
@adamwathan 

This is how the variable is declared:

```scss
:root {
  --color-primary: rgb(255, 0, 0);
}
```

**Reply by afuno** [reply]
Found the problem. It should be done this way:

```scss
:root {
  --color-primary: 255, 0, 0;
}
```

And like this:

```js
primary: ({ opacityVariable }) => `rgba(var(--color-primary), var(${opacityVariable}, 1))`,
```

**Reply by adamwathan** [reply]
Here's a full demo of how to do it correctly:

https://github.com/adamwathan/tailwind-css-variable-text-opacity-demo

**Reply by afuno** [reply]
@adamwathan 

Should this code be exactly like this?

```js
primary: ({ opacityVariable, opacityValue }) => {
  if (opacityValue !== undefined) {
    return `rgba(var(--color-primary), ${opacityValue})`
  }
  if (opacityVariable !== undefined) {
    return `rgba(var(--color-primary), var(${opacityVariable}, 1))`
  }
  return `rgb(var(--color-primary))`
},
```

Everything seems to be successful with this short code:

```js
primary: ({ opacityVariable }) => `rgba(var(--color-primary), var(${opacityVariable}, 1))`
```

🤔

**Reply by adamwathan** [reply]
Yep otherwise it won't work with gradients, which pass an opacityValue rather than an opacityVariable.

**Reply by afuno** [reply]
Okay, got it. Maybe it's better to prepare some function "out of the box" for this?

Like this:

```js
const preparedColor = (colorName, opacityValue, opacityVariable) => {
  if (opacityValue !== undefined) {
    return `rgba(var(${colorName}), ${opacityValue})`
  }

  if (opacityVariable !== undefined) {
    return `rgba(var(${colorName}), var(${opacityVariable}, 1))`
  }

  return `rgb(var(${colorName}))`
}
```
